### PR TITLE
New example: Gradle + Spring Boot + JOOQ, and move everything into /examples/

### PR DESCRIPTION
**Pull request by jord1e**
Thank you for the project, I thought I'd contribute a Gradle example with JOOQ since that's what I am currently using :)

---

@gunnarmorling could you maybe get in contact with the JMC folks to publish the JMC Agent JAR to Maven Central? (Or create a feature request on the bug tracker, I don't have access to it)

https://github.com/openjdk/jmc/blob/master/agent/pom.xml (`org.openjdk.jmc:agent`)

When this is done we can do something cool using the [Gradle distribution plugin](https://docs.gradle.org/current/userguide/distribution_plugin.html):
```kts
plugins {
    distribution
}

val jmcAgent: Configuration by configurations.creating

dependencies {
    jmcAgent("org.openjdk.jmc:agent:8.1.1")
}

distributions {
    create("jmcAgent") {
        distributionBaseName.set("jmc-agent")
        contents {
            from(jmcAgent)
        }
    }
}
```

This creates a task that downloads the JMC Agent jar from Maven central and places it in the build folder, >without third party plugins<
This also gives users the ability to use something like Dependabot or Renovate to automatically update the agent, I already employ this technique for the [OpenTelemetry Javaagent](https://github.com/open-telemetry/opentelemetry-java-instrumentation), and it works beautifully.

Other JMC jars are already being published: https://mvnrepository.com/search?q=org.openjdk.jmc

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Hey, wow, that's really cool, thanks a lot!

> could you maybe get in contact with the JMC folks to publish the JMC Agent JAR to Maven Central? (Or create a feature request on the bug tracker, I don't have access to it)

I spoke to JMC folks about some time ago, but so far it didn't go anywhere yet. I can try and ping them again; but from looking at the code, it seems this example will work just fine for now by downloading the binary provided by Adoptium over HTTP?

**Comment by jord1e** [comment]
Yes the example works perfectly fine! Thank you for the very useful project and your work on MapStruct.

Hopefully you have some better contact with the people running JMC 😄 

Getting the JMC Agent jar published would eliminate the third-party plugin and give Dependabot and Renovate - and the like - a change to update the agent automatically. It would also be able to leverage .m2 and gradle build caching, which may be important for CI environments.

File: spring-jooq-gradle/src/test/java/org/moditect/jfrunit/demos/spring_jooq/SpringJooqGradleApplicationTests.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    @Test
+    public void createAndFindUser() {
+        final String username = String.valueOf(ThreadLocalRandom.current().nextLong());
+        final int age = ThreadLocalRandom.current().nextInt();
+        boolean success = testUserService.createUser(username,
+                age);
+        Assertions.assertThat(success).isTrue();
+
+        jfrEvents.awaitEvents();
+        Assertions.assertThat(jfrEvents.events().filter(this::isQueryEvent).count()).isEqualTo(1);
```
Comment (gunnarmorling) [inline spring-jooq-gradle/src/test/java/org/moditect/jfrunit/demos/spring_jooq/SpringJooqGradleApplicationTests.java:45]: How about adding some assertion on the actual statement as well?

**Review by gunnarmorling — COMMENTED** [review]
Thanks, this looks very good overall. One minor question inline, and one other suggestion: could you add a commit for moving the existing two services (which belong to one example) into a new sub-directory, create a README there with most of the current root-level README and update that one so it essentially just has a sentence or two about the repo's purpose, then links to the two actual examples, and lastly the license?

**Comment by jord1e** [comment]
Will have time for this on Thursday

**Comment by jord1e** [comment]
Done as requested, there is one problem.

We need to capture the first parameter of the following function:
<https://github.com/jOOQ/jOOQ/blob/main/jOOQ/src/main/java/org/jooq/impl/AbstractQuery.java#L410>

I have tried this:
```xml
        <event id="jooq.AbstractQuerySql">
            <label>jOOQ Execute Statement with SQL</label>
            <class>org.jooq.impl.AbstractResultQuery</class>
            <description>Emitted when jOOQ executes a statement, may contain the rendered SQL</description>
            <path>JDBC</path>
            <stacktrace>false</stacktrace>
            <rethrow>false</rethrow>
            <location>WRAP</location>
            <method>
                <name>execute</name>
                <descriptor>(Lorg/jooq/ExecuteContextLorg/jooq/ExecuteListener;)I</descriptor>
                <parameters>
                    <parameter index="0">
                        <name>SQL</name>
                        <description>Rendered SQL String, may be null</description>
                        <contenttype>None</contenttype>
                        <converter>org.moditect.jfrunit.demos.spring_jooq.ExecuteContextSqlConverter</converter>
                    </parameter>
                </parameters>
            </method>
        </event>
```

But it isn't working:
afbeelding

Maybe because it is a protected method? Or can you discover something else?

**Comment by gunnarmorling** [comment]
Thanks a lot for the update! So the reason for not seeing those events was that there was a tiny glitch in the method descriptor: a missing ";" after the first parameter type. I've added that, and also added another event definition for `AbstractResultQuery`. I've removed the `ExecutePreparedStmt` definition as it didn't seem to add much value over the other the. I've also reworked the directory structure bit, so as to have one parent folder for the two Quarkus-related modules. Doing so I goofed up a bit with the .gitignore and committed some Gradle build files. As I didn't want to force-push to this PR, I've created a new one over at https://github.com/moditect/jfrunit-examples/pull/3 which contains everything in the right shape. So I'll close this one and merge the other one. Thanks a lot, @jord1e!

**Comment by jord1e** [comment]
No problem! No. #3 looks great

> So the reason for not seeing those events was that there was a tiny glitch in the method descriptor: a missing ";" after the first parameter type

Urgh, I had it at first but i was doing `(org/jooq/ExecuteContext;org/jooq/ExecuteListener;)I` (Missing `L`)
should have read the spec better 😅

**Comment by gunnarmorling** [comment]
Ha yeah, that syntax isn't super-intuitive indeed.

**Pull request closed without merging**